### PR TITLE
[Implement] CometTask routes for long running tasks; fixes #268

### LINKED ISSUE #268: Improve experience with long running processes wrt clients and predefined timeouts
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the COMET Web Services
- [x] I have searched [open](https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
The current behavior with long tasks is suboptimal. If the task takes more than 60 seconds to process on the server, the clients will often time out with message:  "A task was cancelled" with no further information to the user on the state. Local cache and user interface therefore become desynced from the server until an automatic or manual refresh is done (granted that the task finished meanwhile). This can lead to further downstream errors.

Common examples where this can happen:

- Changes to large Finite State Lists in models where large number of Parameters use them.
- Creation of Engineering Models based on large templates

Solution proposal:

- Investigate the use of 1XX http return codes for progress responses from server (i.e. clients can receive % of tasks complete and messages from server while operation is running).
- Investigate how this can be non-destructively introduced in both server and SDK Dal. Old way should still work.
- Investigate how this can be incorporated in UI to keep user informed properly (loading bars, progress messages). If this method is used, client timeouts should be flexible to support this and not cancel.

### System Configuration

- COMET Web Services version:
  - [ ] CDP4Common:  
  - [ ] PostrgreSQL: 
  - [ ] Other:       
- Environment (Operating system, version and so on):
- .NET Framework version:
- Mono version:
- Additional information:

### PR DESCRIPTION
**Pull request by samatstariongroup**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/COMET-WebServices-Community-Edition/pulls) open
- [x] I have verified that I am following the COMET-SDK [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/COMET-WebServices-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description

Implemented an improvement that for long running POST request a response is returned in the form of a CometTask. The CometTask should be moved to the CDP4-SDK so its a shared class. I propose to do that after this is approved.

### CODE REVIEW

File: CometServer/Modules/10-25/ApiBase.cs [hunk 1]
```diff
@@ -413,14 +473,27 @@ protected async Task WriteJsonResponse(IHeaderInfoProvider headerInfoProvider, I
         /// </returns>
         protected async Task WriteMessagePackResponse(IHeaderInfoProvider headerInfoProvider, IMessagePackSerializer messagePackSerializer, IPermissionInstanceFilterService permissionInstanceFilterService, IReadOnlyList<Thing> resourceResponse, Version requestDataModelVersion, HttpResponse httpResponse, HttpStatusCode statusCode = HttpStatusCode.OK, string requestToken = "")
         {
-            headerInfoProvider.RegisterResponseHeaders(httpResponse, ContentTypeKind.MESSAGEPACK, "");
+            try
+            {
+                var sw = Stopwatch.StartNew();
 
-            httpResponse.StatusCode = (int)statusCode;
+                this.logger.LogInformation("MessagePack serialozation to response started for {token}", requestToken);
```
Comment (antoineatrhea) [inline CometServer/Modules/10-25/ApiBase.cs:480]: typo : serialozation  instead of serialization

**Review by antoineatstariongroup — APPROVED** [review]
Minor editorial stuff found, seams fine for me

File: CometServer/Configuration/LongRunningTasksConfig.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        /// backing field for the <see cref="WaitTime"/> property
+        /// </summary>
+        private int waitTime = 600;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LongRunningTasksConfig"/> class.
+        /// </summary>
+        public LongRunningTasksConfig()
+        {
+            this.retentionTime = 3600;
```
Comment (alexatrhea) [inline CometServer/Configuration/LongRunningTasksConfig.cs:62]: Why reset these values from the above init?

File: CometServer/Modules/10-25/ApiBase.cs [hunk 3]
```diff
@@ -126,15 +148,19 @@ protected ApiBase(IAppConfigService appConfigService, ITokenGeneratorService tok
         /// <param name="appConfigService">
         /// The (injected) <see cref="IAppConfigService"/>
         /// </param>
+        /// <param name="cometHasStartedService">
+        ///
```
Comment (alexatrhea) [inline CometServer/Modules/10-25/ApiBase.cs:152]: docs missing

File: CometServer/Modules/10-25/ApiBase.cs [hunk 4]
```diff
@@ -182,6 +208,24 @@ protected async Task Authorize(IAppConfigService appConfigService, ICredentialsS
...
+        protected async Task<bool> IsServerReady(HttpResponse response)
+        {
+            if (this.CometHasStartedService.GetHasStartedAndIsReady().IsHealthy)
+            {
+                return true;
+            }
+
+            response.ContentType = "application/json";
+            response.StatusCode = (int)HttpStatusCode.ServiceUnavailable;
+            await response.AsJson("not yet started and ready to accept requests");
```
Comment (alexatrhea) [inline CometServer/Modules/10-25/ApiBase.cs:225]: should be "not ready to ..."

**Review by alexatstariongroup — CHANGES_REQUESTED** [review]
Small changes and comment.

**Review by lxatstariongroup — APPROVED** [review]

**Pull request closed without merging**